### DISCUSSION TITLE
Execute parameterized test first in parallel then going sequential

### DISCUSSION THREAD

**Discussion by javanerd** [junit-team/junit-framework#2828] [Q&A]
Hi guys, i just started a request on stackoverflow:
[Request on stackoverflow](https://stackoverflow.com/questions/71070450/junit5-execute-first-parameterized-test-in-parallel-then-going-sequential)
Perhaps someone here has time to look at it. Thank you in advance.

**Comment by portyu9** [comment]
Hi @javanerd 

Today this maps much more naturally to a **parameterized class** than to a single parameterized test method.

`@ParameterizedClass` causes the entire test class to be invoked once for each argument, so your desired structure becomes:

```text
"12345" branch        "67890" branch
     |                     |
 step 1                step 1
     |                     |
 step 2                step 2
     |                     |
 step 3                step 3
     |                     |
 step 4                step 4
```

For current JUnit, for example:

```java
@ParameterizedClass
@ValueSource(strings = {"12345", "67890"})
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
@Execution(ExecutionMode.CONCURRENT)
class TrickyTestExecutionOrder {

    @Parameter
    String number;

    @Test
    @Order(1)
    @Execution(ExecutionMode.SAME_THREAD)
    void firstParallelForEachNumber() {
        // use number
    }

    @Test
    @Order(2)
    @Execution(ExecutionMode.SAME_THREAD)
    void thenDoThis() {
        // use number
    }

    @Test
    @Order(3)
    @Execution(ExecutionMode.SAME_THREAD)
    void andDoThat() {
        // use number
    }

    @Test
    @Order(4)
    @Execution(ExecutionMode.SAME_THREAD)
    void andFinallyThisAgain() {
        // use number
    }
}
```

Enable Jupiter parallel execution, for example in `junit-platform.properties`:

```properties
junit.jupiter.execution.parallel.enabled=true
```

If you specifically want at least two workers:

```properties
junit.jupiter.execution.parallel.config.strategy=fixed
junit.jupiter.execution.parallel.config.fixed.parallelism=2
```

`@Execution(CONCURRENT)` allows the parameterized-class invocations to execute concurrently, while `SAME_THREAD` on the ordered test methods keeps the steps within each invocation on that branch's thread.

Because this example uses `PER_CLASS`, parameterized-class values are injected into the `@Parameter` field. Constructor injection is only supported with the default `PER_METHOD` lifecycle.

One important limitation: `@ParameterizedClass` is still marked **experimental** in current JUnit, so take that into account if your project only permits stable APIs.

If `firstParallelForEachNumber()` is actually setup rather than a separately reportable test, you can also model that part using the parameterized-class lifecycle instead of making it `@Test`.

This capability was not available when the original question was asked in 2022, which is why older solutions are substantially more complicated.

Please mark this answer as accepted if it helped, thank you.
